lasagne/layers/sparse/dense.py

```diff
@@ -0,0 +1,24 @@
+import theano.sparse as sparse
+
+from ..dense import DenseLayer as Dense
+
+
+__all__ = [
+    "DenseLayer",
+]
+
+
+class DenseLayer(Dense):
+    def get_output_for(self, input, **kwargs):
+        """
+        Parameters
+        ----------
+        input : theano sparse matrix
+            output from the previous layer
+        """
+        assert type(input) == sparse.basic.SparseVariable
+
+        activation = sparse.basic.dot(input, self.W)
```

Comment:
> By default, theano.tensor.dot(a,b) will call theano.sparse.basic.structured_dot().

@nouiz , I am not sure about this.

```
>>> import scipy.sparse as sp
>>> import theano
>>> import theano.tensor as T
>>> import theano.sparse as sparse
>>> x = sparse.csr_matrix('x')
>>> y = sparse.csr_matrix('y')
>>> z = T.dot(x, y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/tensor/basic.py", line 4872, in dot
    a, b = as_tensor_variable(a), as_tensor_variable(b)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/tensor/basic.py", line 161, in as_tensor_variable
    "Variable type field must be a TensorType.", x, x.type)
theano.tensor.var.AsTensorError: ('Variable type field must be a TensorType.', SparseVariable{csr,float64}, Sparse[float64, csr])
>>> 
```

> The main difference is in the grad. Just double check you have the gradient behave like you want.

As far as I understood, the `sparse.dot` and `sparse.structured_dot` would only differ, in our case, when computing the grad wrt to the sparse input. As the main propose of these class adaptation are just to allow an sparse input to the network, the grad wrt to the input would not be relevant, since there is no previous layers for learning its parameters.